### PR TITLE
Join Ray Jobs API `JobInfo` with GCS `JobTableData`

### LINKED ISSUE #29621: [jobs] Define a proto API for checking job status
## The Problem

Currently, we have two "overlapping" concepts of jobs:
- GCS Jobs (a Ray driver)
- Submitted jobs (Job submitted to the Ray job server)

There are two ways to check on the status of jobs: using the GCS `ListAllJobs` endpoint (which contains only partial data), and the dashboard list jobs endpoint.

Currently to get full job status from the cluster, you need to query both and join them (based on `gcs_job_table_entry.job_config.metadata["job_submission_id"])`. This has a couple problems:
1. Callers need to hit two separate endpoints, one of which is returning pickled Python bytes (difficult to interpret in other languages).
2. The cluster doesn't fate share with the job server, so it's possible for jobs to still be running even if the job server is down.

## Proposed Solution

1. Have the job server store extended JobInfo in a proto or JSON form, instead of pickled Python data. The proto/JSON JobInfo can still be stored in the internal KV.
2. Enhance `ListAllJobs` to pre-join the data from `job_submission_id` in the internal KV. This means that GCS will need to lookup the extended JobInfo for jobs that have `job_submission_id` defined.
3. Add a `limit` field to ListAllJobs to bound the number of returned jobs.

### PR DESCRIPTION
**Pull request by architkulkarni**
Signed-off-by: Archit Kulkarni <architkulkarni@users.noreply.github.com>

## Why are these changes needed?
- Add a new protobuf for `JobInfo` from the Ray Job API
- Augment the existing GCS `GetAllJobInfo` endpoint to return this information, if available (not all GCS jobs were submitted via the Ray Job API; these jobs won't have this extra `JobInfo`.)

## Related issue number

Closes https://github.com/ray-project/ray/issues/29621
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by architkulkarni** [comment]
@iycheng Could you please help give a preliminary review of the change in the GCS Job manager?

Currently all it's doing is to pull Ray Job API `JobInfo` from the internal KV for each job submitted via the Ray Job API. Because the internal KV get is async, we need some custom logic here involving a counter, but perhaps you know a better pattern to use.  I worked on this with @rickyyx and @edoakes.

What remains to be done in this PR is 
- Actually populate the `JobInfo` into the `GCS_RPC_REPLY` 
- Add tests

File: src/ray/gcs/gcs_server/gcs_job_manager.cc [hunk 1]
```diff
@@ -146,13 +147,66 @@ void GcsJobManager::HandleGetAllJobInfo(rpc::GetAllJobInfoRequest request,
...
-    RAY_LOG(INFO) << "Finished getting all job info.";
-    GCS_RPC_SEND_REPLY(send_reply_callback, reply, Status::OK());
+
+    std::shared_ptr<std::atomic<int>> num_processed_jobs =
+        std::make_shared<std::atomic<int>>(0);
+
+    // For each job with a submission id, we query the internal KV store to get the Ray
+    // Job API JobInfo, and send the gRPC reply in the KV callback. If there are *no* jobs
+    // with a submission id, we must send the gRPC reply in the current function instead.
+    // Use this flag to track it.
```
Comment (edoakes) [inline src/ray/gcs/gcs_server/gcs_job_manager.cc:162]: excellent comment :)

File: src/ray/gcs/gcs_server/gcs_job_manager.cc [hunk 2]
```diff
@@ -146,13 +147,66 @@ void GcsJobManager::HandleGetAllJobInfo(rpc::GetAllJobInfoRequest request,
...
+              reply->mutable_job_info_list(i)->mutable_job_info()->CopyFrom(
+                  jobs_api_info);
+            } else {
+              RAY_LOG(ERROR)
+                  << "Failed to look up Ray Job API JobInfo for job with submission ID "
+                  << job_submission_id;
+            }
+
+            // Send reply if all jobs have been processed.
+            if (num_processed_jobs->fetch_add(1) == reply->job_info_list_size() - 1) {
```
Comment (edoakes) [inline src/ray/gcs/gcs_server/gcs_job_manager.cc:195]: nit: I always find post increments a little hard to follow, might be easier to read this way:
```suggestion
            if (++num_processed_jobs == reply->job_info_list_size()) {
```

Comment (edoakes) [reply src/ray/gcs/gcs_server/gcs_job_manager.cc:195]: note that `x.fetch_add(1)` is equivalent to `x++`

this might be a little uglier given the shared_ptr though... up to you

File: src/ray/gcs/gcs_server/gcs_job_manager.cc [hunk 3]
```diff
@@ -146,13 +147,66 @@ void GcsJobManager::HandleGetAllJobInfo(rpc::GetAllJobInfoRequest request,
...
+              RAY_LOG(INFO) << "Finished getting all job info.";
+              GCS_RPC_SEND_REPLY(send_reply_callback, reply, Status::OK());
+            }
+          };
+
+      std::string job_data_key = "_ray_internal_job_info_" + job_submission_id;
```
Comment (edoakes) [inline src/ray/gcs/gcs_server/gcs_job_manager.cc:201]: should this be a shared constant somewhere?

File: src/ray/gcs/gcs_server/gcs_job_manager.cc [hunk 4]
```diff
@@ -146,13 +147,66 @@ void GcsJobManager::HandleGetAllJobInfo(rpc::GetAllJobInfoRequest request,
...
+      internal_kv_.Get("job", job_data_key, kv_get_callback);
+    }
+
+    // If there were no jobs with a submission id, we must send the reply here.
```
Comment (edoakes) [inline src/ray/gcs/gcs_server/gcs_job_manager.cc:205]:
```suggestion
    // If there were no jobs with a submission id, we must send the reply here because no callbacks will execute.
```

File: src/ray/protobuf/gcs.proto [hunk 5]
```diff
@@ -301,6 +300,37 @@ message GcsNodeInfo {
...
+message JobsAPIInfo {
+  // The status of the job.
+  string status = 1;
+  // The entrypoint command for this job.
+  string entrypoint = 2;
+  // A message describing the status in more detail.
+  string message = 3;
+  // TODO(architkulkarni): Populate this field
+  // Error type (e.g. Runtime env setup failure, Internal error, user script error)
+  string error_type = 4;
```
Comment (edoakes) [inline src/ray/protobuf/gcs.proto:312]: should this be an `enum` rather than a string?

File: src/ray/protobuf/gcs.proto [hunk 6]
```diff
@@ -301,6 +300,37 @@ message GcsNodeInfo {
...
+  // The time when the job was started.  A Unix timestamp in ms.
+  uint64 start_time = 5;
+  // The time when the job moved into a terminal state.  A Unix timestamp in ms.
+  uint64 end_time = 6;
+  // Arbitrary user-provided metadata for the job.
+  map<string, string> metadata = 7;
+  // The JSON-serialized runtime environment for the job.
+  string runtime_env = 8;
```
Comment (edoakes) [inline src/ray/protobuf/gcs.proto:320]:
```suggestion
  // The JSON-serialized runtime environment for the job.
  string runtime_env_json = 8;
```

Comment (edoakes) [reply src/ray/protobuf/gcs.proto:320]: as-is i'd expect it to be a map

**Review by edoakes — COMMENTED** [review]
Looks good, ignore my proto naming suggestions if they aren't helpful

File: src/ray/gcs/gcs_server/gcs_job_manager.cc (hunk 3 above)
Comment (architkulkarni) [reply src/ray/gcs/gcs_server/gcs_job_manager.cc:201]: Yup good catch, I'll add cross-reference comments to the Python

File: src/ray/protobuf/gcs.proto (hunk 5 above)
Comment (architkulkarni) [reply src/ray/protobuf/gcs.proto:312]: Possibly, right now in the Python `JobInfo` it's a string, but it isn't even populated anywhere yet. Unless we need to finalize this decision now (for protobuf backwards compatibility reasons?) I think we can just leave it as string

File: src/ray/gcs/gcs_server/gcs_job_manager.cc (hunk 2 above)
Comment (architkulkarni) [reply src/ray/gcs/gcs_server/gcs_job_manager.cc]: It does seem a little annoying actually:
`src/ray/gcs/gcs_server/gcs_job_manager.cc:195:17: error: cannot increment value of type 'const std::shared_ptr<std::atomic<int> >'
            if (++num_processed_jobs == reply->job_info_list_size()) {`
            
I think I'll keep it as is

File: src/ray/protobuf/gcs.proto (hunk 6 above)
Comment (architkulkarni) [reply src/ray/protobuf/gcs.proto]: Yup that's clearer, fixed now

File: src/ray/protobuf/gcs.proto (hunk 5 above)
Comment (edoakes) [reply src/ray/protobuf/gcs.proto:312]: assuming this is only used internally in ray there's no compatibility issues to worry about

**Review by edoakes — APPROVED** [review]

File: dashboard/modules/job/common.py [hunk 7]
```diff
@@ -127,6 +130,9 @@ def __post_init__(self):
     def to_json(self) -> Dict[str, Any]:
         """Convert this object to a JSON-serializable dictionary.
 
+        Note that the runtime_env field is converted to a JSON-serialized string
```
Comment (rickyyx) [inline dashboard/modules/job/common.py:133]: nit: Possible to keep the naming consistent across the 2? Or we could use a map field for `JobsAPIInfo`?

File: src/ray/gcs/gcs_client/test/global_state_accessor_test.cc [hunk 8]
```diff
@@ -125,6 +125,26 @@ TEST_P(GlobalStateAccessorTest, TestJobTable) {
   ASSERT_EQ(global_state_->GetAllJobInfo().size(), job_count);
 }
 
+// Test GetAllJobInfo where some jobs were submitted by the Ray Job API (i.d. they have
```
Comment (rickyyx) [inline src/ray/gcs/gcs_client/test/global_state_accessor_test.cc:128]: Nice

File: dashboard/modules/job/common.py (hunk 7 above)
Comment (architkulkarni) [reply dashboard/modules/job/common.py:133]: More discussion on the naming here: https://github.com/ray-project/ray/pull/31046#discussion_r1088053459 I don't feel strongly either way.
I think defining a `proto` for RuntimeEnv is out of scope of this PR, it's not a simple map<string, string> unfortunately.

File: src/ray/gcs/gcs_server/gcs_job_manager.cc [hunk 9]
```diff
@@ -146,13 +147,66 @@ void GcsJobManager::HandleGetAllJobInfo(rpc::GetAllJobInfoRequest request,
...
+
+    // For each job with a submission id, we query the internal KV store to get the Ray
+    // Job API JobInfo, and send the gRPC reply in the KV callback. If there are *no* jobs
+    // with a submission id, we must send the gRPC reply in the current function instead.
+    // Use this flag to track it.
+    bool any_jobs_with_submission_id = false;
+
+    for (int i = 0; i < reply->job_info_list_size(); i++) {
+      const auto &metadata = reply->job_info_list(i).config().metadata();
+      if (metadata.find("job_submission_id") == metadata.end()) {
```
Comment (rickyyx) [inline src/ray/gcs/gcs_server/gcs_job_manager.cc:167]: nit: consider saving `find` result to a local iterator to save another lookup at `at` later.

File: src/ray/gcs/gcs_server/gcs_job_manager.cc [hunk 10]
```diff
@@ -146,13 +147,66 @@ void GcsJobManager::HandleGetAllJobInfo(rpc::GetAllJobInfoRequest request,
...
+      any_jobs_with_submission_id = true;
+      const auto &job_submission_id = metadata.at("job_submission_id");
+
+      auto kv_get_callback =
+          [reply, send_reply_callback, num_processed_jobs, i, &job_submission_id](
+              std::optional<std::string> job_info_json) {
+            if (job_info_json) {
+              // Parse the JSON into a JobsAPIInfo proto.
+              rpc::JobsAPIInfo jobs_api_info;
+              RAY_CHECK(google::protobuf::util::JsonStringToMessage(*job_info_json,
```
Comment (rickyyx) [inline src/ray/gcs/gcs_server/gcs_job_manager.cc:181]: There has been past experience where this call will likely dominate the runtime - what's the scale of number of jobs with submission id? And if there's any requirement on the latency of this `GetALLJobInfo`?

Comment (architkulkarni) [reply src/ray/gcs/gcs_server/gcs_job_manager.cc:181]: Not sure, but we should assume it's on the order of the total number of jobs across the lifetime of the cluster (including those without submission_id).  Do we have an estimate of this value? (From observability maybe?)

Not sure about the latency requirement of the endpoint, @ericl any thoughts about this?

**Review by rickyyx — APPROVED** [review]
LGTM! Do we need to enforce a simple `limit` call as well? Or it will be in another PR.

**Comment by architkulkarni** [comment]
There was a pretty basic race condition in the first approach which is now fixed in https://github.com/ray-project/ray/pull/31046/commits/2f57c060d7b1faaef345a44f2ed059c33a2329ae. 

@edoakes @rickyyx mind reviewing it again?

The original approach failed in the case where Job 1/2 had a submission_id, but Job 2/2 had no submission_id.  The reply wouldn't be sent in the kv callback for Job 1 (because not all jobs were processed yet), nor would it be sent after both the jobs were processed (because flag `any_job_with_submission_id` would be True.)


The new approach (which seems simpler and doesn't use the suspicious-looking flag) is to always send the callback when the last job is processed.  @edoakes I don't remember if there was some reason we didn't do this in the first place.

File: src/ray/gcs/gcs_server/gcs_job_manager.cc [hunk 11]
```diff
@@ -197,16 +193,9 @@ void GcsJobManager::HandleGetAllJobInfo(rpc::GetAllJobInfoRequest request,
               GCS_RPC_SEND_REPLY(send_reply_callback, reply, Status::OK());
             }
           };
-
+      RAY_LOG(ERROR) << "Calling internal kv get";
```
Comment (rickyyx) [inline src/ray/gcs/gcs_server/gcs_job_manager.cc]:
```suggestion

```

**Review by rickyyx — APPROVED** [review]

File: src/ray/gcs/gcs_server/gcs_job_manager.cc (hunk 9 above)
Comment (architkulkarni) [reply src/ray/gcs/gcs_server/gcs_job_manager.cc:161]: Done!

**Review by wuisawesome — COMMENTED** [review]
added some non blocking comments

File: src/ray/protobuf/gcs.proto [hunk 12]
```diff
@@ -301,6 +300,39 @@ message GcsNodeInfo {
...
+  string status = 1;
+  // The entrypoint command for this job.
+  string entrypoint = 2;
+  // A message describing the status in more detail.
+  string message = 3;
+  // TODO(architkulkarni): Populate this field
+  // Error type (e.g. Runtime env setup failure, Internal error, user script error)
+  string error_type = 4;
+  // The time when the job was started.  A Unix timestamp in ms.
+  uint64 start_time = 5;
```
[the remark this answers is not in the document]
Comment (architkulkarni) [reply src/ray/protobuf/gcs.proto:316]: Oh good point. I checked and it looks like proto3 does have optional fields (you can check whether they were set or not, and this is different from whether they're 0 or not). So is the right thing to do here to mark them optional in this file? Any downside or risk in doing that?

Also @wuisawesome do you mind reviewing the last commit where a `limit` field is added? That one is `optional` too, which seemed to me like a reasonable way to do it but I couldn't find that pattern used elsewhere so just wanted to double check.  You're codeowner too so if you could approve it that would be appreciated

File: dashboard/modules/job/common.py [hunk 13]
```diff
@@ -61,7 +61,10 @@ def is_terminal(self) -> bool:
 @PublicAPI(stability="stable")
 @dataclass
 class JobInfo:
-    """A class for recording information associated with a job and its execution."""
+    """A class for recording information associated with a job and its execution.
+
+    Please keep this in sync with the JobsAPIInfo proto in src/ray/protobuf/gcs.proto.
```
[the remark this answers is not in the document]
Comment (architkulkarni) [reply dashboard/modules/job/common.py:66]: I think this makes sense for the long term, we can look at it for a future PR. The main issues are converting runtime env to protobuf (it's not a simple dict[str, str] and we have to think about how to make it work with plugins), and we also plan to make JobInfo a pydantic model in the future (I think there's some built in integration with protobuf, but haven't checked the details)

File: src/ray/gcs/gcs_server/gcs_job_manager.cc [hunk 14]
```diff
@@ -146,13 +147,87 @@ void GcsJobManager::HandleGetAllJobInfo(rpc::GetAllJobInfoRequest request,
...
+              std::optional<std::string> job_info_json) {
+            if (job_info_json) {
+              // Parse the JSON into a JobsAPIInfo proto.
+              rpc::JobsAPIInfo jobs_api_info;
+              RAY_CHECK(google::protobuf::util::JsonStringToMessage(*job_info_json,
+                                                                    &jobs_api_info)
+                            .ok());
+              // Load info into the reply.
+              reply->mutable_job_info_list(i)->mutable_job_info()->CopyFrom(
+                  jobs_api_info);
```
Comment (iycheng) [inline src/ray/gcs/gcs_server/gcs_job_manager.cc:216]: nit:
```
*reply->mutable_job_info_list(i)->mutable_job_info() = std::move(jobs_api_info)
```

File: src/ray/gcs/gcs_server/gcs_job_manager.h [hunk 15]
```diff
@@ -24,6 +24,15 @@
 namespace ray {
 namespace gcs {
 
+// Please keep this in sync with the definition in ray_constants.py.
+const std::string kRayInternalNamespacePrefix = "_ray_internal_";
```
Comment (iycheng) [inline src/ray/gcs/gcs_server/gcs_job_manager.h:28]: I think if this file is included twice, it's going to cost duplicate defining of the variables.

These are global variables here.

Comment (iycheng) [reply src/ray/gcs/gcs_server/gcs_job_manager.h:28]: Why not put them into cpp directly if it's not used in other places.

**Comment by iycheng** [comment]
@architkulkarni it seems you want to get everything for a prefix right? Why not just simply extend the internal kv api to have multi-get? It's not difficult, just deligate the call to the StoreClient

https://sourcegraph.com/github.com/ray-project/ray@43a0d8f4cee7ebc763258b4481ac96856798f3fa/-/blob/src/ray/gcs/store_client/store_client.h?L76-79

File: src/ray/gcs/gcs_server/gcs_job_manager.h (hunk 15 above)
Comment (architkulkarni) [reply src/ray/gcs/gcs_server/gcs_job_manager.h:28]: Sure, sounds good

**Comment by architkulkarni** [comment]
> @architkulkarni it seems you want to get everything for a prefix right? Why not just simply extend the internal kv api to have multi-get? It's not difficult, just deligate the call to the StoreClient
> 
> https://sourcegraph.com/github.com/ray-project/ray@43a0d8f4cee7ebc763258b4481ac96856798f3fa/-/blob/src/ray/gcs/store_client/store_client.h?L76-79
> 
> If you have too many jobs there, this is going to be slow because each get will be a rpc to the backend. For memory store, it's fine, but for redis, the request to redis is going to be big.

@iycheng Ah thanks a lot! Let me try to add this.

**Comment by iycheng** [comment]
@architkulkarni sg, let me know if you have any difficulty in doing this especially given the cut is very close.

**Comment by architkulkarni** [comment]
@iycheng You're right, it wasn't too bad. Because it would make the current PR too big, I opted to make a new PR, please review it when you get a chance: https://github.com/ray-project/ray/pull/32096

The current PR will depend on that one.

File: src/ray/protobuf/gcs.proto (hunk 12 above)
[the remark this answers is not in the document]
Comment (architkulkarni) [reply src/ray/protobuf/gcs.proto]: @wuisawesome Made the fields optional, please take a look

**Comment by architkulkarni** [comment]
Failed tests:
- Doc failure "to_summary_by_lineage" unrelated, broken on master
- Datasets `test_execution_optimizer` unrelated, broken on master
- Linkcheck failure "kdnuggets" unrelated and broken on master
- Build images failure "Observed wheel commit () is not expected commit" unrelated and is flaky on master
- `serve:tutorial_rllib` unrelated, broken on master

File: src/ray/gcs/gcs_server/gcs_job_manager.h (hunk 15 above)
Comment (architkulkarni) [reply src/ray/gcs/gcs_server/gcs_job_manager.h:28]: Actually, it's used in a test file, so I think I need to leave it in this `.h` file (I tried moving it to the `.cc` file and the test wouldn't compile).  Is that okay, or is there another workaround?

File: src/ray/gcs/gcs_server/gcs_job_manager.cc (hunk 14 above)
Comment (architkulkarni) [reply src/ray/gcs/gcs_server/gcs_job_manager.cc]: Done

**Review by fishbone — APPROVED** [review]
LG!

File: src/ray/gcs/gcs_server/gcs_job_manager.cc [hunk 16]
```diff
@@ -146,13 +147,76 @@ void GcsJobManager::HandleGetAllJobInfo(rpc::GetAllJobInfoRequest request,
                                         rpc::GetAllJobInfoReply *reply,
                                         rpc::SendReplyCallback send_reply_callback) {
   RAY_LOG(INFO) << "Getting all job info.";
-  auto on_done = [reply, send_reply_callback](
+
+  int limit = std::numeric_limits<int>::max();
```
Comment (iycheng) [inline src/ray/gcs/gcs_server/gcs_job_manager.cc:151]: A more fundamental   problem is that we need pagination for GetAll apis, otherwise, the backend still doing a lot of work.

**Comment by architkulkarni** [comment]
Test failures: 
- rllib multi_agent_independent_learning broken on master, unrelated
- Datasets air:test_tensor_extension was broken on master, unrelated
- Workflows test_basic_workflows_2 flaky on master, unrelated
- serve:tutorial_rllib was broken on master

Java test failed, but it isn't failing on the flaky test tracker. Retrying it to be safe

**Comment by architkulkarni** [comment]
Java test passed, merging
